### DISCUSSION TITLE
Exceeding timeout to openid provider causes auth to not function

### DISCUSSION THREAD

**Discussion by benosman** [directus/directus#14284] [Archive]
### Describe the Bug

This may not a bug with directus per se, but this issue has been causing some headaches.

We use okta for authentication and have two environments, production (okta.com) and staging (oktapreview.com). Recently oktapreview.com has been running a bit slow (on our network at least), and when directus starts up and tries to retrieve the discovery file, it exceeds the timeout set in the openid driver (3500ms).

The issue is that not only is that timeout not configurable, but after the error shows in the log, directus keeps running, without trying to retrieve the discovery file again. This leaves the okta authentication driver broken, and if that is the only driver enabled, it is not possible for anyone to login.

https://github.com/directus/directus/blob/7d451f8df4383373380b7a3eb39d9dfce3488982/api/src/auth/drivers/openid.ts#L53

### To Reproduce

This is difficult to reproduce out the timeout, but our directus instance is configured to have an Okta auth driver using openid, providing client id and secret. It works fine using the production okta instance.

### Errors Shown

RPError: outgoing request timed out after 3500ms
    at /opt/app-root/src/node_modules/openid-client/lib/helpers/request.js:136:13
    at Function.discover (/opt/app-root/src/node_modules/openid-client/lib/issuer.js:171:22)
[ERROR] 17:01:49 RPError: outgoing request timed out after 3500ms

### What version of Directus are you using?

9.12.2

### What version of Node.js are you using?

16.13.0

### What database are you using?

Postgres 12

### What browser are you using?

Chrome, Firefox

### How are you deploying Directus?

Openshift Paas

**Comment by azrikahar** [comment]
Not too familiar with this per se, but just to be sure this is exactly caused by Directus itself, is the "running abit slow" due to lower compute/resource allocated to the staging environment?

However I can definitely see how having the flexibility of [customizing the timeout](https://github.com/panva/node-openid-client/blob/main/docs/README.md#customizing-http-requests) can be nice to have for edge cases like this, and/or handling timeout more gracefully if possible. cc @aidenfoxx for an expert's opinion 🙏

That said, it would also be great if we can somehow verify it is not actually config related issue to make sure we're not fixating on the timeout config as a crutch. 🤔

**Reply by aidenfoxx** [reply]
You're correct. The default `openid-client` timeout is 3500 milliseconds, and we'd need to implement the `setHttpOptionsDefaults` option that you linked to increase that.

The issue occurs when `openid-client` tries to fetch the provider metadata. If it fails an unhandled promise rejection is thrown, and depending on your node version, will terminate Directus.

Myself and Rijk discussed this a while back and we were going to add explicit exception handling and trigger `process.exit(1)` for this. The idea being that if it fails silently it's kinda confusing, since Directus may start but the auth provider will never work (depending on if node terminated or not).

**Reply by benosman** [reply]
@aidenfoxx 

We were using node 14.x and directus didn't terminate we just had the silent fail as you described, we've just switched to 16.x so I'll test how that handles it. Definitely a +1 for me for adding an explicit exception.

Thanks for your work on the auth providers, it's a great help to have them in place, and it made directus use viable for us. One other area we've had some difficulty with though is around the lack of hooks/context for phases of the authentication flow, like handling authentication errors more gracefully with the full okta context for logging (compliance reasons). I've been trying to find a block of time to create an discussion/pull request around this topic but have yet to do so.

**Reply by aidenfoxx** [reply]
I've put up a PR to terminate when we can't fetch the provider metadata https://github.com/directus/directus/pull/14287 now.

We do have the standard auth hooks which do trigger for all auth flows, though they don't provide much provider specific data. There have been some PRs in the past to add extra hooks/data, but I don't know what became of them. Would love to see you pick this up though! 😄

**Comment by aidenfoxx** [comment]
You may also experience this issue if you're running Directus behind a proxy. Without the correct proxy configuration, Directus won't be able to callout to the external Okta URL. See https://github.com/directus/directus/discussions/11387#discussioncomment-2098413 for configuration.

**Comment by ubeytd** [comment]
Our production container just exited because of a failed OpenID request.
